### PR TITLE
GUVNOR-3409: [Guided Decision Table] Change the pattern of an action column

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/GUVNOR-3409

This PR prevents "Conditions" or "BRL Condition Fragments" from being updated if the change means a bound variable used on the RHS becomes unavailable. Similar checks already existed for deleting columns but not updates. I also moved all said checks to the respective _Synchronizers_ for consistency (with how moving columns/rows is prevented should _new state_ be invalid).

### CODE REVIEW

**Comment by jomarko** [comment]
jenkins execute full downstream build

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ActionRetractFactPlugin.java [hunk 1]
```diff
@@ -92,8 +94,13 @@ public Boolean generateColumn() {
         if (isNewColumn()) {
             presenter.appendColumn(editingCol());
         } else {
-            presenter.updateColumn(originalCol(),
-                                   editingCol());
+            try {
+                presenter.updateColumn(originalCol(),
+                                       editingCol());
+            } catch (VetoException veto) {
+                wizard.showGenericVetoError();
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ActionRetractFactPlugin.java]: I saw few `try catch` blocks that just `swallow` the exception, shouldn't they show some message as here? I don't know, just a question.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/commons/BaseDecisionTableColumnPlugin.java [hunk 2]
```diff
@@ -22,6 +22,8 @@
 import org.drools.workbench.models.guided.dtable.shared.model.DTColumnConfig52;
 import org.drools.workbench.models.guided.dtable.shared.model.Pattern52;
 import org.drools.workbench.screens.guided.dtable.client.widget.table.GuidedDecisionTableView;
+import org.drools.workbench.screens.guided.dtable.client.widget.table.model.synchronizers.ModelSynchronizer;
+import org.drools.workbench.screens.guided.dtable.client.widget.table.model.synchronizers.ModelSynchronizer.VetoException;
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/commons/BaseDecisionTableColumnPlugin.java]: Seems like unused imports.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/commons/DecisionTableColumnPlugin.java [hunk 3]
```diff
@@ -21,6 +21,8 @@
 
 import org.drools.workbench.models.guided.dtable.shared.model.DTColumnConfig52;
 import org.drools.workbench.models.guided.dtable.shared.model.Pattern52;
+import org.drools.workbench.screens.guided.dtable.client.widget.table.model.synchronizers.ModelSynchronizer;
+import org.drools.workbench.screens.guided.dtable.client.widget.table.model.synchronizers.ModelSynchronizer.VetoException;
```
Comment (jomarko) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/commons/DecisionTableColumnPlugin.java]: Unused imports

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/commons/BaseDecisionTableColumnPlugin.java (hunk 2 above)
Comment (manstis) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/commons/BaseDecisionTableColumnPlugin.java]: Removed (in next commit!)

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/commons/DecisionTableColumnPlugin.java (hunk 3 above)
Comment (manstis) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/commons/DecisionTableColumnPlugin.java]: Removed (in next commit!)

**Comment by jomarko** [comment]
@manstis here is the scenario from our IRC discussion that should be incorporated:
* Define simple condition like p : Person ( homeAddress : address != null ) - where address is of type Address
* Define BRL action that will change field values of homeAddress
* Try to delete the condition column defined in first step
* Try to update field binding defined in first step

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java [hunk 4]
```diff
@@ -1273,14 +1228,18 @@ public void onDeleteSelectedColumns() {
...
+                    deleteColumn((AttributeCol52) columnToDelete);
+                } else if (columnToDelete instanceof MetadataCol52) {
+                    deleteColumn((MetadataCol52) columnToDelete);
+                } else if (columnToDelete instanceof ConditionCol52) {
+                    deleteColumn((ConditionCol52) columnToDelete);
+                } else if (columnToDelete instanceof ActionCol52) {
+                    deleteColumn((ActionCol52) columnToDelete);
+                }
+            } catch (VetoException veto) {
+                Window.alert("Could not delete!");
```
[the remark this answers is not in the document]
Comment (manstis) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java]: Updated.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/control/ColumnManagementView.java [hunk 5]
```diff
@@ -176,16 +191,28 @@ EditColumnManagementAnchorWidget editAnchor(final ClickHandler clickHandler) {
...
                             });
     }
 
+    void showGenericVetoMessage() {
+        ErrorPopup.showMessage(translate(GuidedDecisionTableErraiConstants.NewGuidedDecisionTableColumnWizard_GenericVetoError));
+    }
+
     void showUnableToDeleteColumnMessage(final ConditionCol52 column) {
-        ErrorPopup.showMessage(GuidedDecisionTableConstants.INSTANCE.UnableToDeleteConditionColumn0(column.getHeader()));
+        ErrorPopup.showMessage(translate(GuidedDecisionTableErraiConstants.NewGuidedDecisionTableColumnWizard_DeletePatternInUseVetoError0,
```
[the remark this answers is not in the document]
Comment (manstis) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/control/ColumnManagementView.java]: No real benefit; other than I chose to use the recommended i18n approach for new development.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ActionRetractFactPlugin.java (hunk 1 above)
Comment (manstis) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ActionRetractFactPlugin.java]: I _know_ as the developer that these scenarios should not happen at the moment (as there are no cases where we'd throw a ```VetoException``` in those scenarios). However I have made them report a generic "veto" message instead - for future proofing.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/control/ColumnManagementView.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jomarko) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/columns/control/ColumnManagementView.java]: Thanks for answer, as you claim is recommended, I will also follow in my PRs.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ActionRetractFactPlugin.java (hunk 1 above)
Comment (jomarko) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/plugins/ActionRetractFactPlugin.java]: Thanks for explanation.

**Comment by manstis** [comment]
@jomarko Updates made.. I was not checking whether a change/deletion in a field was acceptable. I have fixed the use-case you provided and the case when a BRL Condition column is defined with a bound field that is then used in a BRL Action column. Deletion of the BRL Condition should be also be prevented; together with any update to the field binding.

**Comment by jomarko** [comment]
@manstis looks like some tests got broken?

**Comment by manstis** [comment]
@jomarko Tests fixed.

**Pull request merged**